### PR TITLE
C#: Postupdate notes for ternary expressions.

### PR DESCRIPTION
**Pull request by michaelnebel**
This is the C# version of: https://github.com/github/codeql/pull/10444
That is, we can now identify flow out through arguments where expressions are inlined.

Assume we have
```csharp
public class SimpleClass
{
    public string field = "";
}

public void TaintField(SimpleClass sc)
{
    sc.field = "taint source";
}
```
then flow out via `x` and `y` are tracked correctly in the following pseudo example.
```csharp
SimpleClass x = ...
SimpleClass y = ...
TaintField(b ? x : y);
Sink(x.field);
Sink(y.field);
``` 

Furthermore, we also track flow out of expressions like
```csharp
TaintField((SimpleClass)x) // Cast expressions
TaintField(x ?? y) // Null Coalescing expressions
TaintField(x!) // Suppress nullable warning expressions
TaintField(y = x) // Assignment expressions
TaintField(choice switch "x" => x ...)
```

### CODE REVIEW

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 1]
```diff
@@ -177,6 +177,32 @@ predicate hasNodePath(ControlFlowReachabilityConfiguration conf, ExprNode n1, No
   )
 }
 
+/**
+ * Gets a node that may execute last in `n`, and which, when it executes last,
+ * will be the value of `n`.
+ */
+private ControlFlow::Nodes::ExprNode getALastEvalNode(ControlFlow::Nodes::ExprNode cfn) {
+  exists(ConditionalExpr e | cfn.getExpr() = e | result.getExpr() = [e.getThen(), e.getElse()])
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:185]:
```suggestion
  any(LocalExprStepConfiguration x).hasExprPath(_, result, any(ConditionalExpr ce), cfn)
```

Comment (hvitved) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:185]: The reason for this change is that it respects CFG splitting.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -177,6 +177,32 @@ predicate hasNodePath(ControlFlowReachabilityConfiguration conf, ExprNode n1, No
...
+}
+
+private predicate relevantArgumentType(ControlFlow::Nodes::ExprNode cfn) {
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:188]: Perhaps just inline?

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 3]
```diff
@@ -1921,7 +1942,18 @@ private module PostUpdateNodes {
 
     ExprPostUpdateNode() { this = TExprPostUpdateNode(cfn) }
 
-    override ExprNode getPreUpdateNode() { cfn = result.getControlFlowNode() }
+    override ExprNode getPreUpdateNode() {
+      // For compund arguments, such as `m(if b then x else y)`, we want the leaf nodes
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:1946]: `m(b ? x : y)`

**Review by hvitved — CHANGES_REQUESTED** [review]

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 4]
```diff
@@ -410,6 +410,29 @@ module LocalFlow {
...
+  private ControlFlow::Nodes::ExprNode getALastEvalNode(ControlFlow::Nodes::ExprNode cfn) {
+    any(LocalExprStepConfiguration x).hasExprPath(_, result, any(ConditionalExpr ce), cfn)
+  }
+
+  /** Gets a node for which to construct a post-update node for argument `arg`. */
+  ControlFlow::Nodes::ExprNode getAPostUpdateNodeForArg(ControlFlow::Nodes::ExprNode arg) {
+    arg.getExpr() instanceof Argument and
+    result = getALastEvalNode*(arg) and
+    exists(Expr e | result.getExpr() = e |
+      exists(Type t | t = e.stripCasts().getType() |
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:427]: Let's join these two `exists` into one.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll [hunk 5]
```diff
@@ -410,6 +410,27 @@ module LocalFlow {
     n instanceof SummaryNode or
     n instanceof ImplicitCapturedArgumentNode
   }
+
+  /**
+   * Gets a node that may execute last in `n`, and which, when it executes last,
+   * will be the value of `n`.
+   */
+  private ControlFlow::Nodes::ExprNode getALastEvalNode(ControlFlow::Nodes::ExprNode cfn) {
+    any(LocalExprStepConfiguration x).hasExprPath(_, result, any(ConditionalExpr ce), cfn)
```
Comment (hvitved) [inline csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:419]: We should also include
- `NullCoalescingExpr::getAnOperand`
- `SwitchExpr::getACase`
- `SwitchCaseExpr::getBody`

(And ideally also tests.)

**Review by hvitved — CHANGES_REQUESTED** [review]
We are missing some cases, otherwise LGTM.

File: csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll (hunk 5 above)
Comment (michaelnebel) [reply csharp/ql/lib/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll:419]: Alright - I will look into it!

**Review by hvitved — APPROVED** [review]
Thanks, LGTM. Let's see what DCA says before merging.

**Comment by michaelnebel** [comment]
DCA execution has finished.
(1) (Unfortunately) no new alerts were found.
(2) There doesn't appear to be any performance regressions.

**Pull request merged**